File: app/shared/villagers/birthdays.ts
import type { VillagerProfiles } from "@shared/schemas";

// Villager birthdays, on the local calendar. A worktree named after a
// villager celebrates their birthday (Village life), and the name pick
// invites whoever's birthday it is. The date is always passed in, so
// every rule here holds on any day of the year. cli/birthdays.go keeps
// the same rules for the CLI's pick.

const pad = (n: number) => String(n).padStart(2, "0");

// The local calendar day as the profiles spell birthdays, "MM-DD".
export function monthDayOf(date: Date): string {
  return `${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

// A key that changes exactly at local midnight.
export function calendarDayOf(date: Date): string {
  return `${date.getFullYear()}-${monthDayOf(date)}`;
}

function isLeapYear(year: number): boolean {
  return (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;
}

// Whether a "MM-DD" birthday falls on `date`. A leap-day birthday is
// kept on Feb 28 in the three years without a Feb 29.
export function isBirthdayOn(
  birthday: string | undefined,
  date: Date,
): boolean {
  if (birthday === undefined) return false;
  const today = monthDayOf(date);
  if (birthday === today) return true;
  return (
    birthday === "02-29" && today === "02-28" && !isLeapYear(date.getFullYear())
  );
}

// Every character celebrating on `date`, for the name pick to invite.
export function birthdaySlugsOn(
  profiles: VillagerProfiles,
  date: Date,
): string[] {
  return Object.entries(profiles)
    .filter(([, profile]) => isBirthdayOn(profile.birthday, date))
    .map(([slug]) => slug)
    .toSorted();
}

File: cli/birthdays.go
package main

// Villager birthdays, for the name pick: with Village life on, a
// villager whose birthday it is (local date) is invited first. The
// birthdays come from the villager data the app downloads into the data
// dir (app/host/lib/villagers.ts writes villagers/ready/profiles.json
// once a download is whole), so without it there is no one to invite.
// The same rules as the app's shared/villagers/birthdays.ts.

import (
	"path/filepath"
	"sort"
	"time"
)

// Matches villageLifeEnabled in the app's shared/villageLife.ts: only
// with Doubutsu names on, and off unless set.
func villageLifeEnabled(global globalConfig) bool {
	return doubutsuNamesEnabled(global) && global.VillageLife != nil && *global.VillageLife
}

// Whether a "MM-DD" birthday falls on the local calendar day of `day`.
// A leap-day birthday is kept on Feb 28 in years without a Feb 29.
func isBirthdayOn(birthday string, day time.Time) bool {
	today := day.Format("01-02")
	if birthday == today {
		return true
	}
	year := day.Year()
	leap := year%4 == 0 && (year%100 != 0 || year%400 == 0)
	return birthday == "02-29" && today == "02-28" && !leap
}

// The villagers celebrating on `day`, sorted, from the downloaded
// profiles. Empty without Village life or the data: a missing or
// unreadable file just means nobody is invited.
func birthdayGuests(global globalConfig, day time.Time) []string {
	if !villageLifeEnabled(global) {
		return nil
	}
	profiles, ok := readJSONFile[map[string]struct {
		Birthday string `json:"birthday"`
	}](filepath.Join(dataDir(), "villagers", "ready", "profiles.json"))
	if !ok {
		return nil
	}
	var guests []string
	for slug, profile := range profiles {
		if isBirthdayOn(profile.Birthday, day) {
			guests = append(guests, slug)
		}
	}
	sort.Strings(guests)
	return guests
}
